File: euxis-adapters/src/adapters/registry.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2024-2026 Sebastien Rousseau ᛫ https://sebastienrousseau.com

"""Adapter registry for gateway channel configuration."""

from __future__ import annotations

from typing import Any, Dict, Optional

from adapters.slack_adapter import SlackAdapter, SlackAdapterConfig
from adapters.telegram_adapter import TelegramAdapter, TelegramAdapterConfig
from adapters.discord_adapter import DiscordAdapter, DiscordAdapterConfig
from adapters.whatsapp_adapter import WhatsAppAdapter, WhatsAppAdapterConfig
from adapters.sdk import MessageHandler

def build_adapters(config: Dict[str, Any], on_message: Optional[MessageHandler] = None) -> Dict[str, Any]:
    channels = config.get("gateway", {}).get("channels", {})
    adapters: Dict[str, Any] = {}

    slack_cfg = channels.get("slack", {})
    if slack_cfg.get("enabled"):
        adapters["slack"] = SlackAdapter(
            SlackAdapterConfig(
                token=slack_cfg.get("token", ""),
                app_token=slack_cfg.get("app_token", ""),
                mode=slack_cfg.get("mode", "socket"),
            ),
            on_message=on_message,
        )

    telegram_cfg = channels.get("telegram", {})
    if telegram_cfg.get("enabled"):
        adapters["telegram"] = TelegramAdapter(
            TelegramAdapterConfig(
                token=telegram_cfg.get("token", ""),
                mode=telegram_cfg.get("mode", "webhook"),
                webhook_url=telegram_cfg.get("webhook_url", ""),
                poll_timeout=telegram_cfg.get("poll_timeout", 20),
                poll_interval=telegram_cfg.get("poll_interval", 1.5),
            ),
            on_message=on_message,
        )

    discord_cfg = channels.get("discord", {})
    if discord_cfg.get("enabled"):
        adapters["discord"] = DiscordAdapter(
            DiscordAdapterConfig(
                token=discord_cfg.get("token", ""),
                enabled=True,
            ),
            on_message=on_message,
        )

    whatsapp_cfg = channels.get("whatsapp", {})
    if whatsapp_cfg.get("enabled"):
        adapters["whatsapp"] = WhatsAppAdapter(
            WhatsAppAdapterConfig(
                token=whatsapp_cfg.get("token", ""),
                phone_number_id=whatsapp_cfg.get("phone_number_id", ""),
                verify_token=whatsapp_cfg.get("verify_token", ""),
                enabled=True,
            ),
            on_message=on_message,
        )

    return adapters

File: libs/adapters/src/registry.cpp
#include "euxis/adapters/registry.hpp"

namespace euxis::adapters {

auto build_adapters(const nlohmann::json& config, MessageHandler on_message)
    -> std::unordered_map<std::string, AnyAdapter> {
    if (!config.is_object()) {
        return {};
    }
    auto channels =
        config.value("gateway", nlohmann::json::object())
            .value("channels", nlohmann::json::object());

    std::unordered_map<std::string, AnyAdapter> adapters;

    if (auto slack = channels.value("slack", nlohmann::json::object());
        slack.value("enabled", false)) {
        adapters.emplace(std::string("slack"),
                         AnyAdapter(std::make_unique<SlackAdapter>(
                             SlackAdapterConfig{
                                 .token = slack.value("token", ""),
                                 .app_token = slack.value("app_token", ""),
                                 .mode = slack.value("mode", "socket"),
                                 .api_base_url = slack.value("api_base_url", "")},
                             on_message)));
    }

    if (auto tg = channels.value("telegram", nlohmann::json::object());
        tg.value("enabled", false)) {
        adapters.emplace(std::string("telegram"),
                         AnyAdapter(std::make_unique<TelegramAdapter>(
                             TelegramAdapterConfig{
                                 .token = tg.value("token", ""),
                                 .mode = tg.value("mode", "webhook"),
                                 .webhook_url = tg.value("webhook_url", ""),
                                 .poll_timeout = tg.value("poll_timeout", 20),
                                 .poll_interval = tg.value("poll_interval", 1.5),
                                 .api_base_url = tg.value("api_base_url", "")},
                             on_message)));
    }

    if (auto dc = channels.value("discord", nlohmann::json::object());
        dc.value("enabled", false)) {
        adapters.emplace(std::string("discord"),
                         AnyAdapter(std::make_unique<DiscordAdapter>(
                             DiscordAdapterConfig{
                                 .token = dc.value("token", ""),
                                 .enabled = true,
                                 .api_base_url = dc.value("api_base_url", "")},
                             on_message)));
    }

    if (auto wa = channels.value("whatsapp", nlohmann::json::object());
        wa.value("enabled", false)) {
        adapters.emplace(std::string("whatsapp"),
                         AnyAdapter(std::make_unique<WhatsAppAdapter>(
                             WhatsAppAdapterConfig{
                                 .token = wa.value("token", ""),
                                 .phone_number_id = wa.value("phone_number_id", ""),
                                 .verify_token = wa.value("verify_token", ""),
                                 .enabled = true,
                                 .api_base_url = wa.value("api_base_url", "")},
                             on_message)));
    }

    return adapters;
}

} // namespace euxis::adapters
